Added free-threading CPython mode support in MLIR Python bindings

vfdev-5 opened the pull request:
Related to https://github.com/llvm/llvm-project/issues/105522

Description:

This PR is a joint work with Peter Hawkins (@hawkinsp) originally done by myself for pybind11 and then reworked to nanobind based on Peter's branch: https://github.com/hawkinsp/llvm-project/tree/nbdev .

- Added free-threading CPython mode support for MLIR Python bindings
- Added a test which can reveal data races when cpython and LLVM/MLIR compiled with TSAN 

Context:
- Related to https://github.com/google/jax/issues/23073

makslevental commented:
I hit "run tests" but I don't think this is going to work - there are places where we do assume GIL. I don't remember exactly (I believe it's in some of the custom stuff in pybind_adaptors). I can take a closer look soon.

vfdev-5 commented:
@makslevental thanks for the feedback. At the moment, there is no CI job using python 3.13.0rc1 with free-threading enabled, we may 

I run existing python tests on python 3.13 with free-threading and all of them are passing using pybind11 from source (otherwise, there is a deadlock (https://github.com/pybind/pybind11/issues/5346)).


> I don't remember exactly (I believe it's in some of the custom stuff in pybind_adaptors). I can take a closer look soon.

I'll check from my side as well.

Here are few docs on free-threading support in case: 
- https://py-free-threading.github.io/
- https://docs.python.org/3.13/howto/free-threading-extensions.html

vfdev-5 commented:
Set this PR as draft as found non-protected for multi-threading C++ code using `liveContexts` in `PyMlirContext` which is failing in my local multi-threaded tests.

stellaraccident commented:
Thanks for this. I've been burning through free threading deps in other libraries but hadn't come to this one yet. 

Given that this is all very bleeding edge, it may be a little bit before we have proper CI for it upstream, but I'm happy to accept patches that improve free threaded compatibility in the meantime, especially if you have a test rig.

stellaraccident commented:
Maks is right: three are a couple of places where we use mutable global state and we should come up with a convention to protect those with a global mutex. I know there is an idiom for this in CPython itself, but it's there a common thing done for pybind/extensions yet? (Basically where it no ops in gil builds but uses a scoped mutex in free threaded builds)

Let's verify that before marking the modules free threaded compatible.

vfdev-5 commented:
@stellaraccident thanks, i have few questions though about the current python bindings tests using lit.

My local tests are rather simple and I would like to improve that:
```python
import typer

import threading
import concurrent.futures

import mlir.dialects.arith as arith
from mlir.ir import Context, Location, Module, IntegerType, F64Type, InsertionPoint


def mt_run(fn, num_threads, args=(), kwargs={}):
    barrier = threading.Barrier(num_threads)

    def closure():
        barrier.wait()
        return fn(*args, **kwargs)

    with concurrent.futures.ThreadPoolExecutor(
        max_workers=num_threads
    ) as executor:
        futures = []
        for _ in range(num_threads):
            futures.append(executor.submit(closure))
        # We should call future.result() to re-raise an exception if test has
        # failed
        return list(f.result() for f in futures)


def func():
    py_values = [123, 234, 345]
    with Context() as ctx:
        module = Module.create(loc=Location.file("foo.txt", 0, 0))

        dtype = IntegerType.get_signless(64)
        with InsertionPoint(module.body), Location.name("a"):
        # with Location.name("a"):
            arith.constant(dtype, py_values[0])

        with InsertionPoint(module.body), Location.name("b"):
        # with Location.name("b"):
            arith.constant(dtype, py_values[1])

        with InsertionPoint(module.body), Location.name("c"):
        # with Location.name("c"):
            arith.constant(dtype, py_values[2])

    return str(module)


def func2():
    py_values = [123, 234, 345]
    with Context() as ctx, Location.file("foo.txt", 0, 0):
        module = Module.create()
        with InsertionPoint(module.body):
            dtype = IntegerType.get_signless(64)
            arith.constant(dtype, py_values[0])

    return str(module)

def test(func, num_threads=10, expected_first = False):

    if expected_first:
        expected_mlir = func()
        print("Expected MLIR:", expected_mlir)

    output_mlir_list = mt_run(func, num_threads=num_threads)

    if not expected_first:
        expected_mlir = func()
        print("Expected MLIR:", expected_mlir)

    for i, output_mlir in enumerate(output_mlir_list):
        assert output_mlir == expected_mlir, (i, output_mlir, expected_mlir)


def main(
    n: int = 100,
    name: str = "test",
    nt: int = 10,
    ef: bool = False,
):
    test_fn = {
        "test": func,
        "test2": func2,
    }[name]
    for i in range(n):
        print("- Count: ", i)
        test(test_fn, num_threads=nt, expected_first=ef)


if __name__ == "__main__":
    typer.run(main)

```

Ideally, making existing tests to run in a multi-threaded execution (either providing a manual implementation or using tools like: https://github.com/Quansight-Labs/pytest-run-parallel).
Seems like lit is running tests and using stdout output checks which may not always work correctly with multi-treaded execution...

> we should come up with a convention to protect those with a global mutex. I know there is an idiom for this in CPython itself, but is there a common thing done for pybind/extensions yet?

Yes, there is an example in pybind11 for that:
https://github.com/pybind/pybind11/blob/1f8b4a7f1a1c5cc9bd6e0d63fe15540e6c458b24/include/pybind11/detail/internals.h#L645-L649

I applied a similar thing for `getLiveContexts` (locally):
```c++
#ifdef Py_GIL_DISABLED
  static PyMutex &getLock() {
    static PyMutex lock;
    return lock;
  }
#endif

  template<typename F>
  static inline auto withLiveContexts(const F& cb) -> decltype(cb(getLiveContexts())) {
    auto &liveContexts = getLiveContexts();
#ifdef Py_GIL_DISABLED
    auto &lock = getLock();
    PyMutex_Lock(&lock);
#endif
    auto result = cb(liveContexts);
#ifdef Py_GIL_DISABLED
    PyMutex_Unlock(&lock);
#endif
    return result;
  }
```

stellaraccident commented:
That locking idiom looks good to me. Thanks. 

For the tests, I won't claim to love the current testing style, but as with most such things, it is costly to change. If this weren't a part of llvm, I never would have written the tests this way, but I digress...

I suppose we should discuss what level of threading compatibility we are targeting (and likely document that). It seems where this patch is going is ensuring that the library itself does not provide any inherent locking hazards if independent contexts are used from multiple threads. Beyond that, the thread safety is approximately what it is for the underlying mlir libraries, which is to say that it is unsafe to perform mutation on any IR structure that you are unsure as to whether it is being exclusively access by a single thread. Most non framework code will just assume thread per context as the safe way to use the API.

That is quite a bit less safe than the CPython itself, which guards mutable data structures (list, shared dict, etc) internally, making it possible to have logical consistency problems is used across threads in an unsafe way, but internally guarding against corruption of the data structures themselves. 

I'm not aware of a consensus that has emerged on the level of safety that a library such as this should provide. Fine grained locking at the binding layer seems like it would be expensive and error prone, especially given the more relaxed stance of the underlying library. 

Thoughts?

vfdev-5 commented:
Thanks for the reply @stellaraccident !

Yes, we should discuss what level of threading compatibility we are targeting.

> It seems where this patch is going is ensuring that the library itself does not provide any inherent locking hazards if independent contexts are used from multiple threads.

IMO the idea behind enabling CPython free-threading mode in python extensions (like MLIR python bindings) is to make them thread-safe as they are with GIL-enabled CPython.

> I'm not aware of a consensus that has emerged on the level of safety that a library such as this should provide. Fine grained locking at the binding layer seems like it would be expensive and error prone, especially given the more relaxed stance of the underlying library.

Can you please detail this statement? I was thinking that we wanted to make python bindings thread-safe for free-threading cpython...

stellaraccident commented:
> Can you please detail this statement? I was thinking that we wanted to make python bindings thread-safe for free-threading cpython...

If you want to make them thread-safe, then we will need to add much more locking, likely choosing to use context-scoped locks. The issue is that the underlying MLIR library is thread-compatible but not thread-safe. The GIL is currently making usage from Python more thread safe than MLIR actually is. Concretely, think of two threads attempting to modify the operations in the same function as an example of what would happen without more binding-level locking.

Given the bleeding edge nature of free-threaded Python, I don't think we need to get to thread-safe in one step, but we should document and decide on where we are trying to take things. Then we could document a big warning saying that the library is not presently fully thread safe in [list ways].

I'm supportive of making incremental progress in the right direction, so long as we document where we are trying to get to and are clear with folks what the current safety level is.

vfdev-5 commented:
@stellaraccident thanks for the discussion, I agree with your suggestion about documenting the scope of what free-threading would enable.

Concerning this part:

> If you want to make them thread-safe, then we will need to add much more locking, likely choosing to use context-scoped locks. The issue is that the underlying MLIR library is thread-compatible but not thread-safe. The GIL is currently making usage from Python more thread safe than MLIR actually is. Concretely, think of two threads attempting to modify the operations in the same function as an example of what would happen without more binding-level locking.

I would like still to highlight the point that free-threading aims:

>  to retains the thread safety guarantees that were in place before the GIL was removed, and it doesn’t either add nor remove any. So that, for instance, my_int += 1 is unsafe both in the free-threaded and in the default build.
The fact that there is no GIL doesn’t mean that everything must now be thread safe.

(source: https://discuss.python.org/t/free-threading-trove-classifier/62406/24)

A similar situation is now with Numpy which is thread-unsafe to use from both the GIL-enabled and GIL-disabled build and user needs to be careful not to mutate shared state (cc @ngoldbaum). 


> I'm supportive of making incremental progress in the right direction, so long as we document where we are trying to get to and are clear with folks what the current safety level is.

Sounds good, so let's to make MLIR Python bindings tread-safe in free threading mode and document it.

stellaraccident commented:
Yes, but there is an important distinction: `my_int += 1` will not crash in either build. Mutating shared IR state from multiple threads with the gil would produce inconsistent/logical errors but would not crash. Without the gil and without some form of extra locking, it will crash as the underlying IR structures are thread compatible but not thread safe. 

If the cost of fine grain locking is excessive, many projects (mine included) would probably opt to build the extension without locking, since they always use it in thread per context anyway and the IR building is not a user exposed thing.

hawkinsp commented:
I'll also +1 that adding fine-grain locking may be problematic. The MLIR Python builders are already slow, and adding lots of fine grained locking, even uncontended locking, will make that worse. But in JAX, as in Stella's case, we never access the same IR objects from multiple threads, and if we did need to do that we could add client-side locking. I think that's by far the most important mode to support.

stellaraccident commented:
> I'll also +1 that adding fine-grain locking may be problematic. The MLIR Python builders are already slow, and adding lots of fine grained locking, even uncontended locking, will make that worse. But in JAX, as in Stella's case, we never access the same IR objects from multiple threads, and if we did need to do that we could add client-side locking. I think that's by far the most important mode to support.

Off topic, but one of these days, I'm going to have to create a stripped down, build only (probably nanobind based) fast version of the API. Not at the top of the list right now, but yeah... slow. On the torch side, we only use the non-ODS API so that we have more flexibility here in the future.

hawkinsp commented:
> Off topic, but one of these days, I'm going to have to create a stripped down, build only (probably nanobind based) fast version of the API. Not at the top of the list right now, but yeah... slow. On the torch side, we only use the non-ODS API so that we have more flexibility here in the future.

Yeah, at some point soon I'm either going to end up trying to send PRs to speed up the existing bindings (switch to nanobind?) or I'll migrate off them and do my own thing.

stellaraccident commented:
> > Off topic, but one of these days, I'm going to have to create a stripped down, build only (probably nanobind based) fast version of the API. Not at the top of the list right now, but yeah... slow. On the torch side, we only use the non-ODS API so that we have more flexibility here in the future.
> 
> Yeah, at some point soon I'm either going to end up trying to send PRs to speed up the existing bindings (switch to nanobind?) or I'll migrate off them and do my own thing.

I haven't benchmarked it in detail, but there are quite a bit of "productivity features" in there. I don't think for fastest-path builder use, the general IR manipulation APIs are going to be super easy to upgrade -- and they do get use. If the whole thing was ported to nanobind, I'd feel better about contributing a build-only fast-path subset library that could live alongside. In the current state, that would go out of tree because if going to the trouble to write something fast, I'm not going to base it on pybind.

hawkinsp committed:
Enabled freethreading support in MLIR python bindings

WIP freethreading

Added free-threading CPython mode support in Python bindings
- temporarily updated requirements

Added lock on PyGlobals::get and PyMlirContext liveContexts
WIP on adding multithreaded_tests

More tests and added a lock to _cext.register_operation

Updated tests + labelled passing and failing tests

Updated locks and added docs

Docs updates, tests and fixed cmake config issue

Fixed nanobind target in mlir/cmake/modules/AddMLIRPython.cmake

Reverted mlir/test/python/lib/PythonTestModulePybind11.cpp

Removed old data-races stacks and removed fixed tests

Revert some python deps versions

Recoded mlir/test/python/multithreaded_tests.py without pytest

Added a condition to skip running mlir/test/python/multithreaded_tests.py when has GIL

Addressed some of the PR review comments

Updated Python docs and mlir/test/python/multithreaded_tests.py

Updated mlir/test/python/multithreaded_tests.py due to removed lock in ExecutionEngine

hawkinsp commented:
I have a branch in which I've ported @vfdev-5 's PR for free-threading on top of nanobind: https://github.com/llvm/llvm-project/compare/main...hawkinsp:llvm-project:nbdev

I'm happy to send that out for review, but it depends on the other nanobind PRs I have open.

stellaraccident commented:
Thanks. It would be good to get this landed. I won't have time myself to have a close look until after the holiday. But if you get someone else who knows this stuff to review/land, that would work for me too.

hawkinsp commented:
I've been testing free threading in JAX and found that this something very close to this change (see my branch above) was sufficient to eliminate all the TSAN errors I saw from MLIR Python bindings, outside things that I believe are false positives or intentionally not thread-safe (e.g., IR dumping).

I'd propose we land and iterate? I don't think any of the substantive changes here are likely to be surprising or controversial, and while they may not fix all possible races they certainly fix enough to be useful. They should also have little downside if you aren't enabling free-threading, since, e.g. `nb::ft_lock_guard` expands to nothing at all if free-threading isn't enabled.

I suspect the test here may be the most interesting piece to discuss (it uses pytest, which is new, and is best run under TSAN, which you may or may not have in your CI infrastructure for MLIR Python), but if that is going to prove a matter for a long debate I'd suggest iterating on it in a separate PR.

hawkinsp commented on mlir/test/python/lib/PythonTestModulePybind11.cpp:
```diff
@@ -23,7 +23,7 @@ static bool mlirTypeIsARankedIntegerTensor(MlirType t) {
          mlirTypeIsAInteger(mlirShapedTypeGetElementType(t));
 }
 
-PYBIND11_MODULE(_mlirPythonTestPybind11, m) {
```
Why this change? I think this was correct as is.

vfdev-5 replied:
Thanks for catching that! this is due to an incorrect rebasing

hawkinsp commented on mlir/test/python/multithreaded_tests.py:
````diff
@@ -0,0 +1,669 @@
+```
+E               WARNING: ThreadSanitizer: data race (pid=19669)
+E                 Write of size 8 at 0x7f3882591840 by thread T62:
+E                   #0 unicode_fill_utf8 /tmp/cpython-tsan/Objects/unicodeobject.c:5444:30 (python3.13t+0x3142a1) (BuildId: 31ded8fac9cea6b44dd4f3246be3853edec8d0dc)
````
This is a CPython bug (https://github.com/python/cpython/issues/128013) which is now fixed at head. Link that here?

(Also your stack traces need updating for nanobind).

vfdev-5 replied:
The fix was backported to 3.13: https://github.com/python/cpython/pull/128417 so we can retest and remove this

hawkinsp commented on mlir/test/python/multithreaded_tests.py:
````diff
@@ -0,0 +1,669 @@
+
+LLVM related data-races
+1) mlir pass manager
+```
````
Almost all of these races are the same race, which is that `llvm::raw_fd_ostream` is not thread-safe. That appears to be an intentional decision. I'm not sure if the MLIR Python bindings should deviate from this or not; I leave that up to someone who knows this code more.

It might be sufficient to document this behavior.

hawkinsp commented on mlir/test/python/multithreaded_tests.py:
````diff
@@ -0,0 +1,669 @@
+1) ctypes
+```
+WARNING: ThreadSanitizer: data race (pid=99593)
+  Atomic read of size 1 at 0x7f6054c485a8 by thread T3:
````
Your belief is that this is a CPython/libffi bug or false positive, correct? Did we ever establish that either way?

vfdev-5 replied:
Finally, maybe, it is not a false positive. I could build both libffi and cpython with TSAN and simple test shows data race, more info here: https://gist.github.com/vfdev-5/06b1c60713241e75773e9a92f2ecf135

stellaraccident commented:
> I've been testing free threading in JAX and found that this something very close to this change (see my branch above) was sufficient to eliminate all the TSAN errors I saw from MLIR Python bindings, outside things that I believe are false positives or intentionally not thread-safe (e.g., IR dumping).
> 
> I'd propose we land and iterate? I don't think any of the substantive changes here are likely to be surprising or controversial, and while they may not fix all possible races they certainly fix enough to be useful. They should also have little downside if you aren't enabling free-threading, since, e.g. `nb::ft_lock_guard` expands to nothing at all if free-threading isn't enabled.
> 
> I suspect the test here may be the most interesting piece to discuss (it uses pytest, which is new, and is best run under TSAN, which you may or may not have in your CI infrastructure for MLIR Python), but if that is going to prove a matter for a long debate I'd suggest iterating on it in a separate PR.

Works for me. Iirc, I got hung up on this needing a pybind version bump and then just got too busy and forgot. But your upgrade to nanobind handled the compatibility issue. 

Once landed, one of us should update the docs to call out the level status and known pointy bits. Not blocking, though.

vfdev-5 commented:
> I suspect the test here may be the most interesting piece to discuss (it uses pytest, which is new, and is best run under TSAN, which you may or may not have in your CI infrastructure for MLIR Python), but if that is going to prove a matter for a long debate I'd suggest iterating on it in a separate PR.

I updated the tests such that it now uses unittest and only runs with free-threaded cpython.

hawkinsp commented on mlir/CMakeLists.txt:
```diff
@@ -170,7 +170,7 @@ configure_file(
 #   The python executable is correct (set with -DPython3_EXECUTABLE=...)
 # By default, find_package and probing for installed pybind11 is performed.
-# Super projects can set MLIR_DISABLE_CONFIGURE_PYTHON_DEV_PACKAGES=ON to 
+# Super projects can set MLIR_DISABLE_CONFIGURE_PYTHON_DEV_PACKAGES=ON to
```
I'm not familiar with LLVM's review processes, but if it were my code base, I'd tell you "please make unrelated changes in a separate PR". It's trivial to review a PR that, say, fixes doc comments or typos, and it helps when reviewing and landing changes to keep PRs focused on one thing and one thing only. And it when looking, at, say, line-by-line blame to know whether a change came from a substantive PR or a trivial cleanup PR.

vfdev-5 replied:
Yes, I agree, this is a good practice in OSS
https://github.com/llvm/llvm-project/pull/121791

hawkinsp commented on mlir/docs/Bindings/Python.md:
````diff
@@ -1035,7 +1035,7 @@ class ConstantOp(_ods_ir.OpView):
 ```
 
-expects `value` to be a `TypedAttr` (e.g., `IntegerAttr` or `FloatAttr`). 
+expects `value` to be a `TypedAttr` (e.g., `IntegerAttr` or `FloatAttr`).
````
e.g., here also.

Same everywhere.

vfdev-5 replied:
I exported these changes here: https://github.com/llvm/llvm-project/pull/121791, so I update this PR once the other one with cosmetics will be merged and these changes will disappear.
My IDE automaticallly will remove trailing space etc so prefer to wait a bit...

hawkinsp commented on mlir/docs/Bindings/Python.md:
```diff
@@ -1181,9 +1181,50 @@ make the passes available along with the dialect.
+
+## Free-threading (No-GIL) support
+
+Free-threading or no-GIL support refers to CPython interpreter (>=3.13) with Global Interpreter Lock made optional. For details on the topic, please check [PEP-703](https://peps.python.org/pep-0703/) and this [link](https://py-free-threading.github.io/).
```
Nit: it's better to say something like "[the Python free-threading guide]" as the link anchor, rather than "this link".

hawkinsp commented on mlir/docs/Bindings/Python.md:
```diff
@@ -1181,9 +1181,50 @@ make the passes available along with the dialect.
+
+Free-threading or no-GIL support refers to CPython interpreter (>=3.13) with Global Interpreter Lock made optional. For details on the topic, please check [PEP-703](https://peps.python.org/pep-0703/) and this [link](https://py-free-threading.github.io/).
+
+MLIR Python bindings are made free-threading compatible with exceptions (discussed below) in the following sense: it is safe to work in multiple threads with **independent** contexts/modules. Below we show an example code of safe usage:
```
Nit: "The MLIR Python bindings are free-threading compatible" (remove made).

I suspect we should say just one of "contexts" or "modules" here, not both. "contexts" is certainly true and is the more conservative thing to say. It may also be true of modules, but I'm less sure; we'd need tests that made concurrent mutations to different modules in the same context.

hawkinsp commented on mlir/docs/Bindings/Python.md:
````diff
@@ -1181,9 +1181,50 @@ make the passes available along with the dialect.
+```
+
+The exceptions to the free-threading compatibility:
+- Usage of Python `ctypes` is unsafe
````
I might skip this, because it's actually a CPython/libffi bug (https://github.com/python/cpython/issues/128485), likely benign, and in general we don't want to document all known bugs in some CPython version here.

Keep the docs focused on limitations in MLIR.

vfdev-5 replied:
Actually, by unsafe ctypes I aggregated a list of problems, one of which is about ctypes/libffi. 
I'll make again the list of them and report on cpython.

hawkinsp commented on mlir/docs/Bindings/Python.md:
```diff
@@ -1181,9 +1181,50 @@ make the passes available along with the dialect.
+The exceptions to the free-threading compatibility:
+- Usage of Python `ctypes` is unsafe
+- IR printing is unsafe, e.g. when using `PassManager` with `PassManager.enable_ir_printing()` which calls thread-unsafe `llvm::raw_ostream`.
+- Usage of `mlirEmitError`, `mlirOperationDump` is unsafe
```
Use the Python API names for these?

vfdev-5 replied:
These are actually again about unsafe raw_ostream:
```
- Usage of `Location.emit_error` which calls thread-unsafe `llvm::raw_ostream`.
- Usage of `Module.dump` which calls thread-unsafe `llvm::raw_ostream`.
```

hawkinsp commented on mlir/docs/Bindings/Python.md:
```diff
@@ -1181,9 +1181,50 @@ make the passes available along with the dialect.
+- Usage of `mlirEmitError`, `mlirOperationDump` is unsafe
+
+
+For more details, please see the list of xfailed tests in `mlir/test/python/multithreaded_tests.py`.
```
I might skip this as too much detail.

hawkinsp commented on mlir/examples/standalone/python/StandaloneExtensionPybind11.cpp:
```diff
@@ -12,9 +12,11 @@
 using namespace mlir::python::adaptors;
 
-PYBIND11_MODULE(_standaloneDialectsPybind11, m) {
+PYBIND11_MODULE(_standaloneDialects, m, py::mod_gil_not_used()) {
```
Again I think this should be `_standaloneDialectsPybind11`, because we're building both pybind11 and nanobind versions of the same module.

hawkinsp commented on mlir/lib/Bindings/Python/Globals.h:
```diff
@@ -124,6 +134,7 @@ class PyGlobals {
   /// Set of dialect namespaces that we have attempted to import implementation
   /// modules for.
   llvm::StringSet<> loadedDialectModules;
+
```
Nit: revert this newline.

hawkinsp commented on mlir/test/python/execution_engine.py:
```diff
@@ -306,7 +306,7 @@ def callback(a):
 
     with Context():
-        # The module takes a subview of the argument memref, casts it to an unranked memref and 
+        # The module takes a subview of the argument memref, casts it to an unranked memref and
```
Revert.

hawkinsp commented on mlir/lib/Bindings/Python/IRCore.cpp:
```diff
@@ -263,22 +270,28 @@ struct PyGlobalDebugFlag {
     return PyGlobals::get().lookupAttributeBuilder(attributeKind).has_value();
   }
-  static nb::callable dundeGetItemNamed(const std::string &attributeKind) {
+  static nb::callable dunderGetItemNamed(const std::string &attributeKind) {
```
Make this change in another PR.

vfdev-5 replied:
Done here: https://github.com/llvm/llvm-project/pull/121794

hawkinsp commented on mlir/test/python/multithreaded_tests.py:
```diff
@@ -0,0 +1,526 @@
+Observed warnings reported by TSAN.
+
+CPython and free-threading known data-races:
+1) ctypes, libffi, dlmalloc
```
Link the CPython issue.

hawkinsp commented on mlir/test/python/multithreaded_tests.py:
````diff
@@ -0,0 +1,526 @@
+    #3 PyCFuncPtr_new /project/cpython/./Modules/_ctypes/_ctypes.c:3955:13 (_ctypes.cpython-313t-x86_64-linux-gnu.so+0x18ba4) (BuildId: e9fba41204b37973fe60945270803301c7df9615)
+```
+
+2) LLVM related data-races, llvm::raw_ostream is not thread-safe
````
Well, we should *not* observe these, right? The test should only test things that don't exhibit them?

(This is a great reason not to run those test cases!)

vfdev-5 commented on mlir/lib/Bindings/Python/ExecutionEngineModule.cpp:
```diff
@@ -26,6 +26,8 @@ class PyExecutionEngine {
     other.executionEngine.ptr = nullptr;
   }
   ~PyExecutionEngine() {
+    // Avoid races in llvm::orc::LLJIT::deinitialize
```
I added another lock here to prevent a race in a bunch of tests using ExecutionEngine and having races with a stack similar to
```
==================
WARNING: ThreadSanitizer: data race (pid=243501)
  Write of size 8 at 0x724000001628 by thread T47:
    #0 memmove <null> (python3.13t+0xd9c2f) (BuildId: 65e52348ca9319985f9f5e265fd43aab2c867d59)
    #1 <null> <null> (libgcc_s.so.1+0x243c3) (BuildId: 3072445288dd2aba348bf583c65f70509aab8141)
    #2 llvm::RuntimeDyldELF::registerEHFrames() /project/llvm-project/llvm/lib/ExecutionEngine/RuntimeDyld/RuntimeDyldELF.cpp:226:12 (libMLIRPythonCAPI.so.20.0git+0x924f25c) (BuildId: 2ae0848dae8d0b3b)
    #3 llvm::RuntimeDyldImpl::finalizeAsync(std::unique_ptr<llvm::RuntimeDyldImpl, std::default_delete<llvm::RuntimeDyldImpl>>, llvm::unique_function<void (llvm::object::OwningBinary<llvm::object::ObjectFile>, std::unique_ptr<llvm::RuntimeDyld::LoadedObjectInfo, std::default_delete<llvm::RuntimeDyld::LoadedObjectInfo>>, llvm::Error)>, llvm::object::OwningBinary<llvm::object::ObjectFile>, std::unique_ptr<llvm::RuntimeDyld::LoadedObjectInfo, std::default_delete<llvm::RuntimeDyld::LoadedObjectInfo>>)::$_0::operator()(llvm::Expected<std::map<llvm::StringRef, llvm::JITEvaluatedSymbol, std::less<llvm::StringRef>, std::allocator<std::pair<llvm::StringRef const, llvm::JITEvaluatedSymbol>>>>) /project/llvm-project/llvm/lib/ExecutionEngine/RuntimeDyld/RuntimeDyld.cpp:1256:21 (libMLIRPythonCAPI.so.20.0git+0x923c952) (BuildId: 2ae0848dae8d0b3b)
    #4 void llvm::detail::UniqueFunctionBase<void, llvm::Expected<std::map<llvm::StringRef, llvm::JITEvaluatedSymbol, std::less<llvm::StringRef>, std::allocator<std::pair<llvm::StringRef const, llvm::JITEvaluatedSymbol>>>>>::CallImpl<llvm::RuntimeDyldImpl::finalizeAsync(std::unique_ptr<llvm::RuntimeDyldImpl, std::default_delete<llvm::RuntimeDyldImpl>>, llvm::unique_function<void (llvm::object::OwningBinary<llvm::object::ObjectFile>, std::unique_ptr<llvm::RuntimeDyld::LoadedObjectInfo, std::default_delete<llvm::RuntimeDyld::LoadedObjectInfo>>, llvm::Error)>, llvm::object::OwningBinary<llvm::object::ObjectFile>, std::unique_ptr<llvm::RuntimeDyld::LoadedObjectInfo, std::default_delete<llvm::RuntimeDyld::LoadedObjectInfo>>)::$_0>(void*, llvm::Expected<std::map<llvm::StringRef, llvm::JITEvaluatedSymbol, std::less<llvm::StringRef>, std::allocator<std::pair<llvm::StringRef const, llvm::JITEvaluatedSymbol>>>>&) /project/llvm-project/llvm/include/llvm/ADT/FunctionExtras.h:222:12 (libMLIRPythonCAPI.so.20.0git+0x9242d48) (BuildId: 2ae0848dae8d0b3b)
    #5 llvm::unique_function<void (llvm::Expected<std::map<llvm::StringRef, llvm::JITEvaluatedSymbol, std::less<llvm::StringRef>, std::allocator<std::pair<llvm::StringRef const, llvm::JITEvaluatedSymbol>>>>)>::operator()(llvm::Expected<std::map<llvm::StringRef, llvm::JITEvaluatedSymbol, std::less<llvm::StringRef>, std::allocator<std::pair<llvm::StringRef const, llvm::JITEvaluatedSymbol>>>>) /project/llvm-project/llvm/include/llvm/ADT/FunctionExtras.h:387:12 (libMLIRPythonCAPI.so.20.0git+0x9208087) (BuildId: 2ae0848dae8d0b3b)
    #6 (anonymous namespace)::JITDylibSearchOrderResolver::lookup(std::set<llvm::StringRef, std::less<llvm::StringRef>, std::allocator<llvm::StringRef>> const&, llvm::unique_function<void (llvm::Expected<std::map<llvm::StringRef, llvm::JITEvaluatedSymbol, std::less<llvm::StringRef>, std::allocator<std::pair<llvm::StringRef const, llvm::JITEvaluatedSymbol>>>>)>)::'lambda'(llvm::Expected<llvm::DenseMap<llvm::orc::SymbolStringPtr, llvm::orc::ExecutorSymbolDef, llvm::DenseMapInfo<llvm::orc::SymbolStringPtr, void>, llvm::detail::DenseMapPair<llvm::orc::SymbolStringPtr, llvm::orc::ExecutorSymbolDef>>>)::operator()(llvm::Expected<llvm::DenseMap<llvm::orc::SymbolStringPtr, llvm::orc::ExecutorSymbolDef, llvm::DenseMapInfo<llvm::orc::SymbolStringPtr, void>, llvm::detail::DenseMapPair<llvm::orc::SymbolStringPtr, llvm::orc::ExecutorSymbolDef>>>) /project/llvm-project/llvm/lib/ExecutionEngine/Orc/RTDyldObjectLinkingLayer.cpp:47:11 (libMLIRPythonCAPI.so.20.0git+0x9208087)
    #7 void llvm::detail::UniqueFunctionBase<void, llvm::Expected<llvm::DenseMap<llvm::orc::SymbolStringPtr, llvm::orc::ExecutorSymbolDef, llvm::DenseMapInfo<llvm::orc::SymbolStringPtr, void>, llvm::detail::DenseMapPair<llvm::orc::SymbolStringPtr, llvm::orc::ExecutorSymbolDef>>>>::CallImpl<(anonymous namespace)::JITDylibSearchOrderResolver::lookup(std::set<llvm::StringRef, std::less<llvm::StringRef>, std::allocator<llvm::StringRef>> const&, llvm::unique_function<void (llvm::Expected<std::map<llvm::StringRef, llvm::JITEvaluatedSymbol, std::less<llvm::StringRef>, std::allocator<std::pair<llvm::StringRef const, llvm::JITEvaluatedSymbol>>>>)>)::'lambda'(llvm::Expected<llvm::DenseMap<llvm::orc::SymbolStringPtr, llvm::orc::ExecutorSymbolDef, llvm::DenseMapInfo<llvm::orc::SymbolStringPtr, void>, llvm::detail::DenseMapPair<llvm::orc::SymbolStringPtr, llvm::orc::ExecutorSymbolDef>>>)>(void*, llvm::Expected<llvm::DenseMap<llvm::orc::SymbolStringPtr, llvm::orc::ExecutorSymbolDef, llvm::DenseMapInfo<llvm::orc::SymbolStringPtr, void>, llvm::detail::DenseMapPair<llvm::orc::SymbolStringPtr, llvm::orc::ExecutorSymbolDef>>>&) /project/llvm-project/llvm/include/llvm/ADT/FunctionExtras.h:222:12 (libMLIRPythonCAPI.so.20.0git+0x9208087)
    #8 llvm::unique_function<void (llvm::Expected<llvm::DenseMap<llvm::orc::SymbolStringPtr, llvm::orc::ExecutorSymbolDef, llvm::DenseMapInfo<llvm::orc::SymbolStringPtr, void>, llvm::detail::DenseMapPair<llvm::orc::SymbolStringPtr, llvm::orc::ExecutorSymbolDef>>>)>::operator()(llvm::Expected<llvm::DenseMap<llvm::orc::SymbolStringPtr, llvm::orc::ExecutorSymbolDef, llvm::DenseMapInfo<llvm::orc::SymbolStringPtr, void>, llvm::detail::DenseMapPair<llvm::orc::SymbolStringPtr, llvm::orc::ExecutorSymbolDef>>>) /project/llvm-project/llvm/include/llvm/ADT/FunctionExtras.h:387:12 (libMLIRPythonCAPI.so.20.0git+0x90ed255) (BuildId: 2ae0848dae8d0b3b)
    #9 llvm::orc::AsynchronousSymbolQuery::handleComplete(llvm::orc::ExecutionSession&)::RunQueryCompleteTask::run() /project/llvm-project/llvm/lib/ExecutionEngine/Orc/Core.cpp:246:27 (libMLIRPythonCAPI.so.20.0git+0x90ed255)
    #10 llvm::orc::InPlaceTaskDispatcher::dispatch(std::unique_ptr<llvm::orc::Task, std::default_delete<llvm::orc::Task>>) /project/llvm-project/llvm/lib/ExecutionEngine/Orc/TaskDispatch.cpp:23:68 (libMLIRPythonCAPI.so.20.0git+0x92154ad) (BuildId: 2ae0848dae8d0b3b)
    #11 llvm::orc::ExecutionSession::dispatchTask(std::unique_ptr<llvm::orc::Task, std::default_delete<llvm::orc::Task>>) /project/llvm-project/llvm/include/llvm/ExecutionEngine/Orc/Core.h:1550:26 (libMLIRPythonCAPI.so.20.0git+0x90a4fc5) (BuildId: 2ae0848dae8d0b3b)
    #12 llvm::orc::AsynchronousSymbolQuery::handleComplete(llvm::orc::ExecutionSession&) /project/llvm-project/llvm/lib/ExecutionEngine/Orc/Core.cpp:256:6 (libMLIRPythonCAPI.so.20.0git+0x90a4e80) (BuildId: 2ae0848dae8d0b3b)
    #13 llvm::orc::JITDylib::resolve(llvm::orc::MaterializationResponsibility&, llvm::DenseMap<llvm::orc::SymbolStringPtr, llvm::orc::ExecutorSymbolDef, llvm::DenseMapInfo<llvm::orc::SymbolStringPtr, void>, llvm::detail::DenseMapPair<llvm::orc::SymbolStringPtr, llvm::orc::ExecutorSymbolDef>> const&) /project/llvm-project/llvm/lib/ExecutionEngine/Orc/Core.cpp:975:8 (libMLIRPythonCAPI.so.20.0git+0x90b45c4) (BuildId: 2ae0848dae8d0b3b)
    #14 llvm::orc::ExecutionSession::OL_notifyResolved(llvm::orc::MaterializationResponsibility&, llvm::DenseMap<llvm::orc::SymbolStringPtr, llvm::orc::ExecutorSymbolDef, llvm::DenseMapInfo<llvm::orc::SymbolStringPtr, void>, llvm::detail::DenseMapPair<llvm::orc::SymbolStringPtr, llvm::orc::ExecutorSymbolDef>> const&) /project/llvm-project/llvm/lib/ExecutionEngine/Orc/Core.cpp:2915:16 (libMLIRPythonCAPI.so.20.0git+0x90d14cc) (BuildId: 2ae0848dae8d0b3b)
    #15 llvm::orc::MaterializationResponsibility::notifyResolved(llvm::DenseMap<llvm::orc::SymbolStringPtr, llvm::orc::ExecutorSymbolDef, llvm::DenseMapInfo<llvm::orc::SymbolStringPtr, void>, llvm::detail::DenseMapPair<llvm::orc::SymbolStringPtr, llvm::orc::ExecutorSymbolDef>> const&) /project/llvm-project/llvm/include/llvm/ExecutionEngine/Orc/Core.h:1944:32 (libMLIRPythonCAPI.so.20.0git+0x909e6da) (BuildId: 2ae0848dae8d0b3b)
    #16 llvm::orc::AbsoluteSymbolsMaterializationUnit::materialize(std::unique_ptr<llvm::orc::MaterializationResponsibility, std::default_delete<llvm::orc::MaterializationResponsibility>>) /project/llvm-project/llvm/lib/ExecutionEngine/Orc/AbsoluteSymbols.cpp:31:21 (libMLIRPythonCAPI.so.20.0git+0x909e6da)
    #17 llvm::orc::MaterializationTask::run() /project/llvm-project/llvm/lib/ExecutionEngine/Orc/Core.cpp:1584:39 (libMLIRPythonCAPI.so.20.0git+0x90c3002) (BuildId: 2ae0848dae8d0b3b)
    #18 llvm::orc::InPlaceTaskDispatcher::dispatch(std::unique_ptr<llvm::orc::Task, std::default_delete<llvm::orc::Task>>) /project/llvm-project/llvm/lib/ExecutionEngine/Orc/TaskDispatch.cpp:23:68 (libMLIRPythonCAPI.so.20.0git+0x92154ad) (BuildId: 2ae0848dae8d0b3b)
    #19 llvm::orc::ExecutionSession::dispatchTask(std::unique_ptr<llvm::orc::Task, std::default_delete<llvm::orc::Task>>) /project/llvm-project/llvm/include/llvm/ExecutionEngine/Orc/Core.h:1550:26 (libMLIRPythonCAPI.so.20.0git+0x90a4fc5) (BuildId: 2ae0848dae8d0b3b)
    #20 llvm::orc::ExecutionSession::dispatchOutstandingMUs() /project/llvm-project/llvm/lib/ExecutionEngine/Orc/Core.cpp:2174:5 (libMLIRPythonCAPI.so.20.0git+0x90c661b) (BuildId: 2ae0848dae8d0b3b)
    #21 llvm::orc::ExecutionSession::OL_completeLookup(std::unique_ptr<llvm::orc::InProgressLookupState, std::default_delete<llvm::orc::InProgressLookupState>>, std::shared_ptr<llvm::orc::AsynchronousSymbolQuery>, std::function<void (llvm::DenseMap<llvm::orc::JITDylib*, llvm::DenseSet<llvm::orc::SymbolStringPtr, llvm::DenseMapInfo<llvm::orc::SymbolStringPtr, void>>, llvm::DenseMapInfo<llvm::orc::JITDylib*, void>, llvm::detail::DenseMapPair<llvm::orc::JITDylib*, llvm::DenseSet<llvm::orc::SymbolStringPtr, llvm::DenseMapInfo<llvm::orc::SymbolStringPtr, void>>>> const&)>) /project/llvm-project/llvm/lib/ExecutionEngine/Orc/Core.cpp:2793:3 (libMLIRPythonCAPI.so.20.0git+0x90cf2fd) (BuildId: 2ae0848dae8d0b3b)
    #22 llvm::orc::InProgressFullLookupState::complete(std::unique_ptr<llvm::orc::InProgressLookupState, std::default_delete<llvm::orc::InProgressLookupState>>) /project/llvm-project/llvm/lib/ExecutionEngine/Orc/Core.cpp:578:8 (libMLIRPythonCAPI.so.20.0git+0x90f89d8) (BuildId: 2ae0848dae8d0b3b)
    #23 llvm::orc::ExecutionSession::OL_applyQueryPhase1(std::unique_ptr<llvm::orc::InProgressLookupState, std::default_delete<llvm::orc::InProgressLookupState>>, llvm::Error) /project/llvm-project/llvm/lib/ExecutionEngine/Orc/Core.cpp:2541:11 (libMLIRPythonCAPI.so.20.0git+0x90ae6b4) (BuildId: 2ae0848dae8d0b3b)
    #24 llvm::orc::ExecutionSession::lookup(llvm::orc::LookupKind, std::vector<std::pair<llvm::orc::JITDylib*, llvm::orc::JITDylibLookupFlags>, std::allocator<std::pair<llvm::orc::JITDylib*, llvm::orc::JITDylibLookupFlags>>> const&, llvm::orc::SymbolLookupSet, llvm::orc::SymbolState, llvm::unique_function<void (llvm::Expected<llvm::DenseMap<llvm::orc::SymbolStringPtr, llvm::orc::ExecutorSymbolDef, llvm::DenseMapInfo<llvm::orc::SymbolStringPtr, void>, llvm::detail::DenseMapPair<llvm::orc::SymbolStringPtr, llvm::orc::ExecutorSymbolDef>>>)>, std::function<void (llvm::DenseMap<llvm::orc::JITDylib*, llvm::DenseSet<llvm::orc::SymbolStringPtr, llvm::DenseMapInfo<llvm::orc::SymbolStringPtr, void>>, llvm::DenseMapInfo<llvm::orc::JITDylib*, void>, llvm::detail::DenseMapPair<llvm::orc::JITDylib*, llvm::DenseSet<llvm::orc::SymbolStringPtr, llvm::DenseMapInfo<llvm::orc::SymbolStringPtr, void>>>> const&)>) /project/llvm-project/llvm/lib/ExecutionEngine/Orc/Core.cpp:1814:3 (libMLIRPythonCAPI.so.20.0git+0x90a41ab) (BuildId: 2ae0848dae8d0b3b)
    #25 (anonymous namespace)::JITDylibSearchOrderResolver::lookup(std::set<llvm::StringRef, std::less<llvm::StringRef>, std::allocator<llvm::StringRef>> const&, llvm::unique_function<void (llvm::Expected<std::map<llvm::StringRef, llvm::JITEvaluatedSymbol, std::less<llvm::StringRef>, std::allocator<std::pair<llvm::StringRef const, llvm::JITEvaluatedSymbol>>>>)>) /project/llvm-project/llvm/lib/ExecutionEngine/Orc/RTDyldObjectLinkingLayer.cpp:53:8 (libMLIRPythonCAPI.so.20.0git+0x9207423) (BuildId: 2ae0848dae8d0b3b)
    #26 llvm::RuntimeDyldImpl::finalizeAsync(std::unique_ptr<llvm::RuntimeDyldImpl, std::default_delete<llvm::RuntimeDyldImpl>>, llvm::unique_function<void (llvm::object::OwningBinary<llvm::object::ObjectFile>, std::unique_ptr<llvm::RuntimeDyld::LoadedObjectInfo, std::default_delete<llvm::RuntimeDyld::LoadedObjectInfo>>, llvm::Error)>, llvm::object::OwningBinary<llvm::object::ObjectFile>, std::unique_ptr<llvm::RuntimeDyld::LoadedObjectInfo, std::default_delete<llvm::RuntimeDyld::LoadedObjectInfo>>) /project/llvm-project/llvm/lib/ExecutionEngine/RuntimeDyld/RuntimeDyld.cpp:1279:26 (libMLIRPythonCAPI.so.20.0git+0x923bee5) (BuildId: 2ae0848dae8d0b3b)
    #27 llvm::jitLinkForORC(llvm::object::OwningBinary<llvm::object::ObjectFile>, llvm::RuntimeDyld::MemoryManager&, llvm::JITSymbolResolver&, bool, llvm::unique_function<llvm::Error (llvm::object::ObjectFile const&, llvm::RuntimeDyld::LoadedObjectInfo&, std::map<llvm::StringRef, llvm::JITEvaluatedSymbol, std::less<llvm::StringRef>, std::allocator<std::pair<llvm::StringRef const, llvm::JITEvaluatedSymbol>>>)>, llvm::unique_function<void (llvm::object::OwningBinary<llvm::object::ObjectFile>, std::unique_ptr<llvm::RuntimeDyld::LoadedObjectInfo, std::default_delete<llvm::RuntimeDyld::LoadedObjectInfo>>, llvm::Error)>) /project/llvm-project/llvm/lib/ExecutionEngine/RuntimeDyld/RuntimeDyld.cpp:1488:3 (libMLIRPythonCAPI.so.20.0git+0x923ef65) (BuildId: 2ae0848dae8d0b3b)
    #28 llvm::orc::RTDyldObjectLinkingLayer::emit(std::unique_ptr<llvm::orc::MaterializationResponsibility, std::default_delete<llvm::orc::MaterializationResponsibility>>, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>) /project/llvm-project/llvm/lib/ExecutionEngine/Orc/RTDyldObjectLinkingLayer.cpp:193:3 (libMLIRPythonCAPI.so.20.0git+0x9201a11) (BuildId: 2ae0848dae8d0b3b)
    #29 llvm::orc::ObjectTransformLayer::emit(std::unique_ptr<llvm::orc::MaterializationResponsibility, std::default_delete<llvm::orc::MaterializationResponsibility>>, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>) /project/llvm-project/llvm/lib/ExecutionEngine/Orc/ObjectTransformLayer.cpp:40:13 (libMLIRPythonCAPI.so.20.0git+0x91fdc35) (BuildId: 2ae0848dae8d0b3b)
    #30 llvm::orc::IRCompileLayer::emit(std::unique_ptr<llvm::orc::MaterializationResponsibility, std::default_delete<llvm::orc::MaterializationResponsibility>>, llvm::orc::ThreadSafeModule) /project/llvm-project/llvm/lib/ExecutionEngine/Orc/IRCompileLayer.cpp:40:15 (libMLIRPythonCAPI.so.20.0git+0x911e2d2) (BuildId: 2ae0848dae8d0b3b)
    #31 llvm::orc::IRTransformLayer::emit(std::unique_ptr<llvm::orc::MaterializationResponsibility, std::default_delete<llvm::orc::MaterializationResponsibility>>, llvm::orc::ThreadSafeModule) /project/llvm-project/llvm/lib/ExecutionEngine/Orc/IRTransformLayer.cpp:24:15 (libMLIRPythonCAPI.so.20.0git+0x9149f2f) (BuildId: 2ae0848dae8d0b3b)
    #32 llvm::orc::IRTransformLayer::emit(std::unique_ptr<llvm::orc::MaterializationResponsibility, std::default_delete<llvm::orc::MaterializationResponsibility>>, llvm::orc::ThreadSafeModule) /project/llvm-project/llvm/lib/ExecutionEngine/Orc/IRTransformLayer.cpp:24:15 (libMLIRPythonCAPI.so.20.0git+0x9149f2f) (BuildId: 2ae0848dae8d0b3b)
    #33 llvm::orc::BasicIRLayerMaterializationUnit::materialize(std::unique_ptr<llvm::orc::MaterializationResponsibility, std::default_delete<llvm::orc::MaterializationResponsibility>>) /project/llvm-project/llvm/lib/ExecutionEngine/Orc/Layer.cpp:158:5 (libMLIRPythonCAPI.so.20.0git+0x912331d) (BuildId: 2ae0848dae8d0b3b)
    #34 llvm::orc::MaterializationTask::run() /project/llvm-project/llvm/lib/ExecutionEngine/Orc/Core.cpp:1584:39 (libMLIRPythonCAPI.so.20.0git+0x90c3002) (BuildId: 2ae0848dae8d0b3b)
    #35 llvm::orc::InPlaceTaskDispatcher::dispatch(std::unique_ptr<llvm::orc::Task, std::default_delete<llvm::orc::Task>>) /project/llvm-project/llvm/lib/ExecutionEngine/Orc/TaskDispatch.cpp:23:68 (libMLIRPythonCAPI.so.20.0git+0x92154ad) (BuildId: 2ae0848dae8d0b3b)
    #36 llvm::orc::ExecutionSession::dispatchTask(std::unique_ptr<llvm::orc::Task, std::default_delete<llvm::orc::Task>>) /project/llvm-project/llvm/include/llvm/ExecutionEngine/Orc/Core.h:1550:26 (libMLIRPythonCAPI.so.20.0git+0x90a4fc5) (BuildId: 2ae0848dae8d0b3b)
    #37 llvm::orc::ExecutionSession::dispatchOutstandingMUs() /project/llvm-project/llvm/lib/ExecutionEngine/Orc/Core.cpp:2174:5 (libMLIRPythonCAPI.so.20.0git+0x90c661b) (BuildId: 2ae0848dae8d0b3b)
    #38 llvm::orc::ExecutionSession::OL_completeLookup(std::unique_ptr<llvm::orc::InProgressLookupState, std::default_delete<llvm::orc::InProgressLookupState>>, std::shared_ptr<llvm::orc::AsynchronousSymbolQuery>, std::function<void (llvm::DenseMap<llvm::orc::JITDylib*, llvm::DenseSet<llvm::orc::SymbolStringPtr, llvm::DenseMapInfo<llvm::orc::SymbolStringPtr, void>>, llvm::DenseMapInfo<llvm::orc::JITDylib*, void>, llvm::detail::DenseMapPair<llvm::orc::JITDylib*, llvm::DenseSet<llvm::orc::SymbolStringPtr, llvm::DenseMapInfo<llvm::orc::SymbolStringPtr, void>>>> const&)>) /project/llvm-project/llvm/lib/ExecutionEngine/Orc/Core.cpp:2793:3 (libMLIRPythonCAPI.so.20.0git+0x90cf2fd) (BuildId: 2ae0848dae8d0b3b)
    #39 llvm::orc::InProgressFullLookupState::complete(std::unique_ptr<llvm::orc::InProgressLookupState, std::default_delete<llvm::orc::InProgressLookupState>>) /project/llvm-project/llvm/lib/ExecutionEngine/Orc/Core.cpp:578:8 (libMLIRPythonCAPI.so.20.0git+0x90f89d8) (BuildId: 2ae0848dae8d0b3b)
    #40 llvm::orc::ExecutionSession::OL_applyQueryPhase1(std::unique_ptr<llvm::orc::InProgressLookupState, std::default_delete<llvm::orc::InProgressLookupState>>, llvm::Error) /project/llvm-project/llvm/lib/ExecutionEngine/Orc/Core.cpp:2541:11 (libMLIRPythonCAPI.so.20.0git+0x90ae6b4) (BuildId: 2ae0848dae8d0b3b)
    #41 llvm::orc::ExecutionSession::lookup(llvm::orc::LookupKind, std::vector<std::pair<llvm::orc::JITDylib*, llvm::orc::JITDylibLookupFlags>, std::allocator<std::pair<llvm::orc::JITDylib*, llvm::orc::JITDylibLookupFlags>>> const&, llvm::orc::SymbolLookupSet, llvm::orc::SymbolState, llvm::unique_function<void (llvm::Expected<llvm::DenseMap<llvm::orc::SymbolStringPtr, llvm::orc::ExecutorSymbolDef, llvm::DenseMapInfo<llvm::orc::SymbolStringPtr, void>, llvm::detail::DenseMapPair<llvm::orc::SymbolStringPtr, llvm::orc::ExecutorSymbolDef>>>)>, std::function<void (llvm::DenseMap<llvm::orc::JITDylib*, llvm::DenseSet<llvm::orc::SymbolStringPtr, llvm::DenseMapInfo<llvm::orc::SymbolStringPtr, void>>, llvm::DenseMapInfo<llvm::orc::JITDylib*, void>, llvm::detail::DenseMapPair<llvm::orc::JITDylib*, llvm::DenseSet<llvm::orc::SymbolStringPtr, llvm::DenseMapInfo<llvm::orc::SymbolStringPtr, void>>>> const&)>) /project/llvm-project/llvm/lib/ExecutionEngine/Orc/Core.cpp:1814:3 (libMLIRPythonCAPI.so.20.0git+0x90a41ab) (BuildId: 2ae0848dae8d0b3b)
    #42 llvm::orc::Platform::lookupInitSymbols(llvm::orc::ExecutionSession&, llvm::DenseMap<llvm::orc::JITDylib*, llvm::orc::SymbolLookupSet, llvm::DenseMapInfo<llvm::orc::JITDylib*, void>, llvm::detail::DenseMapPair<llvm::orc::JITDylib*, llvm::orc::SymbolLookupSet>> const&) /project/llvm-project/llvm/lib/ExecutionEngine/Orc/Core.cpp:1506:8 (libMLIRPythonCAPI.so.20.0git+0x90c1d3c) (BuildId: 2ae0848dae8d0b3b)
    #43 (anonymous namespace)::GenericLLVMIRPlatformSupport::getDeinitializers(llvm::orc::JITDylib&) /project/llvm-project/llvm/lib/ExecutionEngine/Orc/LLJIT.cpp:377:25 (libMLIRPythonCAPI.so.20.0git+0x913e5b5) (BuildId: 2ae0848dae8d0b3b)
    #44 (anonymous namespace)::GenericLLVMIRPlatformSupport::deinitialize(llvm::orc::JITDylib&) /project/llvm-project/llvm/lib/ExecutionEngine/Orc/LLJIT.cpp:255:31 (libMLIRPythonCAPI.so.20.0git+0x913e5b5)
    #45 llvm::orc::LLJIT::deinitialize(llvm::orc::JITDylib&) /project/llvm-project/llvm/include/llvm/ExecutionEngine/Orc/LLJIT.h:212:16 (libMLIRPythonCAPI.so.20.0git+0x5d8dc76) (BuildId: 2ae0848dae8d0b3b)
    #46 mlir::ExecutionEngine::~ExecutionEngine() /project/llvm-project/mlir/lib/ExecutionEngine/ExecutionEngine.cpp:226:29 (libMLIRPythonCAPI.so.20.0git+0x5d8d8d3) (BuildId: 2ae0848dae8d0b3b)
    #47 mlirExecutionEngineDestroy /project/llvm-project/mlir/lib/CAPI/ExecutionEngine/ExecutionEngine.cpp:72:3 (libMLIRPythonCAPI.so.20.0git+0x46ad01e) (BuildId: 2ae0848dae8d0b3b)
    #48 (anonymous namespace)::PyExecutionEngine::~PyExecutionEngine() /project/llvm-project/mlir/lib/Bindings/Python/ExecutionEngineModule.cpp:30:7 (_mlirExecutionEngine.cpython-313t-x86_64-linux-gnu.so+0xc0d8) (BuildId: 44fbde6f86fbff75)

```

hawkinsp replied:
Hmm. I think this "race" is actually a false positive.

`registerEHFrames` ultimately calls `__register_frame` from `libgcc`. I believe the `__register_frame` from `libgcc` is thread safe via locking, if built with threading enabled. But we didn't instrument `libgcc` with tsan instrumentation, so it's not surprising that tsan doesn't propagate happens-before information through it.

I think I might revert this change and add a tsan suppression.

vfdev-5 replied:
I suppose that you are talking about this code: 
https://github.com/gcc-mirror/gcc/blob/0c28cf56fe97bfd0ee4110ad2b8cac33bd513447/libgcc/unwind-dw2-fde.c#L143-L153

vfdev-5 replied:
Confirm that this is a false-positive when tested with libgcc_s built with TSAN

hawkinsp commented:
To me, this looks ready for a review from one of the MLIR Python owners.

jpienaar commented on mlir/test/python/multithreaded_tests.py:
```diff
@@ -0,0 +1,518 @@
+    with Context(), Location.unknown():
+        module = Module.create()
+        with InsertionPoint(module.body):
+            print("\nTEST:", f.__name__)
```
Is print inside rather than at top of function as in others here on purpose?

jpienaar commented on mlir/test/python/multithreaded_tests.py:
```diff
@@ -0,0 +1,518 @@
+    return f
+
+
+def run_with_insertion_point_v4(f):
```
Could you perhaps add a comment or choose more self-documenting function name? (when reading v2 vs v3, I wouldn't be able to tell what it is doing below)

vfdev-5 replied:
I can also add in comments from which test module the run function comes from.

jpienaar commented on mlir/test/python/multithreaded_tests.py:
```diff
@@ -0,0 +1,518 @@
+        output_folder = Path(test_cls.output_folder.name)
+
+        for test_mod_info in test_modules:
+            assert isinstance(test_mod_info, tuple) and len(test_mod_info) in (2, 3)
```
Could you document these constants?

vfdev-5 replied:
These are the current tests implementation details:
- 2 is the tuple's length, for example `("ir/affine_expr", run)` and 
- 3 is when we would like to include test functions found in the test module which does not have the name as `test*`:
```python
    (
        "dialects/transform_interpreter",
        run_with_context_and_location,
        ["print_", "transform_options", "failed", "include"],
    ),
```
I can add comments

jpienaar reviewed (approved):
LG, thanks! The test section could have been good with less manual lists - but that's probably as much on us currently in how tests are written than anything else. Also this increases testing/comfort in free-threading a lot.

ScottTodd commented on mlir/cmake/modules/AddMLIRPython.cmake:
```diff
@@ -668,12 +668,31 @@ function(add_mlir_python_extension libname extname)
     if (LLVM_COMPILER_IS_GCC_COMPATIBLE OR CLANG_CL)
       # Avoids warnings from upstream nanobind.
-      target_compile_options(nanobind-static
+      set(nanobind_target "nanobind-static")
+      if (NOT TARGET ${nanobind_target})
+        # Get correct nanobind target name: nanobind-static-ft or something else
+        # It is set by nanobind_add_module function according to the passed options
+        get_property(all_targets DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR} PROPERTY BUILDSYSTEM_TARGETS)
```
I'm seeing issues with this code when building in the downstream https://github.com/iree-org/iree project, using Python 3.13t. In my project, the `add_mlir_python_extension` function here is running from a different CMake source directory that does not include _any_ `nanobind` targets, as they are defined in other source files:

Logs with some extra debugging messages:
```
  --   Building iree-dialects project at /home/nod/dev/projects/iree/llvm-external-projects/iree-dialects (into /home/nod/dev/projects/iree/compiler/build/b/llvm-external-projects/mlir-iree-dialects)
  --   libname IREEDialectsPythonModules.extension._mlirRegisterEverything.dso using 'nanobind'
  --   all_targets in /home/nod/dev/projects/iree/llvm-external-projects/iree-dialects/python: IREEDialectsPythonSources;IREEDialectsPythonExtensions;IREEDialectsPythonSources.Dialects;IREEDialectsPythonSources.Dialects.iree_input;IREEDialectsPythonSources.Dialects.iree_input.tablegen;IREEDialectsPythonSources.Dialects.iree_input.ops_gen;IREEDialectsPythonSources.Dialects.iree_structured_transform;IREEDialectsPythonSources.Dialects.iree_structured_transform.tablegen;IREEDialectsPythonSources.Dialects.iree_structured_transform.ops_gen;IREEDialectsPythonExtensions.Main;IREEDialectsAggregateCAPI;IREEDialectsAggregateCAPI.sources;IREEDialectsPythonModules;IREEDialectsPythonModules.extension._mlirRegisterEverything.dso
  CMake Error at compiler/build/b/lib/cmake/mlir/AddMLIRPython.cmake:696 (message):
    Could not find nanobind target to set compile options to
  Call Stack (most recent call first):
    compiler/build/b/lib/cmake/mlir/AddMLIRPython.cmake:235 (add_mlir_python_extension)
    compiler/build/b/lib/cmake/mlir/AddMLIRPython.cmake:256 (_process_target)
    llvm-external-projects/iree-dialects/python/CMakeLists.txt:82 (add_mlir_python_modules)
```

full logs from CI: https://github.com/iree-org/iree/actions/runs/12901589475/job/35973909192#step:12:53035

The target name with 3.13t is `nanobind-static-ft`, but this search misses it because of the different CMake source.

---

I can think of a few solutions, if this target lookup and options setting is truly required:

1. Search through an explicit list of target names using `if (TARGET)`
2. Iterate over _all_ targets in the project, not just the targets in the current directory, using code like https://stackoverflow.com/a/62311397
3. Iterate over targets in the directory known to MLIR (`llvm-project/mlir/python`)

ScottTodd replied:
My downstream project is also setting `MLIR_DISABLE_CONFIGURE_PYTHON_DEV_PACKAGES` to take control over the Python dependency setup. Another solution is to only modify the target options if `MLIR_DISABLE_CONFIGURE_PYTHON_DEV_PACKAGES` is not set.

hawkinsp replied:
CMake is mysterious to me, but would https://github.com/wjakob/nanobind/pull/868 help here? All we're trying to do here is suppress the warnings.

ScottTodd replied:
Maybe... I don't see warnings from nanobind when building in IREE, but IREE does set its own compile options and it imports using FetchContent instead of find_package.

Proposing a change/fix: https://github.com/llvm/llvm-project/pull/123997
